taiga2/controllers/models_controller.py

```diff
@@ -1300,6 +1300,36 @@ def get_datafile_by_taiga_id(taiga_id: str, one_or_none=False) -> Optional[DataF
     return resolve_virtual_datafile(datafile)
 
 
+def log_datafile_read_access_info(datafile_id: int):
+    # If the user has read this data file, a row with this datafiles access info will already exist, so update that
+    # row's access count and last access time info. Otherwise, add a row to track the read access of this file for this user.
+    user = get_current_session_user()
+    user_id = user.id
+
+    read_access_row = (
```

Comment:
Ah! This is exciting! This is a perfect example of something that is _super_ common when working with databases, and it looks correct ... *but* it has a flaw which may cause an exception to be thrown randomly. 

This is a great thing to learn about! Let me see if I can find someone who has written about this...

Reply:
Actually, to start, let me describe the scenario under which you'd see an error.

Imagine a user, Bob, kicks off a distribute job which runs on 100 machines in parallel. At the start of the job, it fetches a file from Taiga. 95 jobs complete fine, but 5 of the jobs failed. Upon looking at the logs, it appears that line 1330 has thrown an exception saying that an insert failed because the primary key already exists.

So, the mystery is: how is this possible?

Reply:
bonus question: If Bob re-runs his 100 jobs, will he be able to recreate this issue?

Reply:
(Quickly looking around, the best resource I have found is this SO post: https://stackoverflow.com/questions/7165998/how-to-do-an-upsert-with-sqlalchemy. The specific problem I wanted to flag is described in https://stackoverflow.com/a/46838022 )